File: fc-chat/demo/part3/ch3/03-2-message-system-add-channel/src/main/java/net/prostars/messagesystem/handler/websocket/WriteMessageHandler.java
package net.prostars.messagesystem.handler.websocket;

import net.prostars.messagesystem.constant.IdKey;
import net.prostars.messagesystem.dto.domain.ChannelId;
import net.prostars.messagesystem.dto.domain.UserId;
import net.prostars.messagesystem.dto.websocket.inbound.WriteMessage;
import net.prostars.messagesystem.dto.websocket.outbound.MessageNotification;
import net.prostars.messagesystem.service.MessageService;
import net.prostars.messagesystem.service.UserService;
import net.prostars.messagesystem.session.WebSocketSessionManager;
import org.springframework.stereotype.Component;
import org.springframework.web.socket.WebSocketSession;

@Component
@SuppressWarnings("unused")
public class WriteMessageHandler implements BaseRequestHandler<WriteMessage> {

  private final UserService userService;
  private final MessageService messageService;
  private final WebSocketSessionManager webSocketSessionManager;

  public WriteMessageHandler(
      UserService userService,
      MessageService messageService,
      WebSocketSessionManager webSocketSessionManager) {
    this.userService = userService;
    this.messageService = messageService;
    this.webSocketSessionManager = webSocketSessionManager;
  }

  @Override
  public void handleRequest(WebSocketSession senderSession, WriteMessage request) {
    UserId senderUserId = (UserId) senderSession.getAttributes().get(IdKey.USER_ID.getValue());
    ChannelId channelId = request.getChannelId();
    String content = request.getContent();
    String senderUsername = userService.getUsername(senderUserId).orElse("unknown");
    messageService.sendMessage(
        senderUserId,
        content,
        channelId,
        (participantId) -> {
          WebSocketSession participantSession = webSocketSessionManager.getSession(participantId);
          MessageNotification messageNotification =
              new MessageNotification(channelId, senderUsername, content);
          if (participantSession != null) {
            webSocketSessionManager.sendMessage(participantSession, messageNotification);
          }
        });
  }
}

File: fc-chat/demo/part3/with-push/src/main/kotlin/com/yscorp/withpush/messagesystem/handler/websocket/WriteMessageHandler.kt
package com.yscorp.withpush.messagesystem.handler.websocket

import com.yscorp.withpush.messagesystem.constant.IdKey
import com.yscorp.withpush.messagesystem.dto.domain.UserId
import com.yscorp.withpush.messagesystem.dto.websocket.inbound.WriteMessage
import com.yscorp.withpush.messagesystem.dto.websocket.outbound.MessageNotification
import com.yscorp.withpush.messagesystem.service.MessageService
import com.yscorp.withpush.messagesystem.service.UserService
import org.springframework.stereotype.Component
import org.springframework.web.socket.WebSocketSession

@Component
@Suppress("unused")
class WriteMessageHandler(
    private val userService: UserService,
    private val messageService: MessageService
) : BaseRequestHandler<WriteMessage> {

    /**
     * 클라이언트로부터 메시지 전송 요청을 처리합니다.
     * 사용자 정보와 메시지를 받아 메시지 서비스를 통해 전송하고, 알림도 생성합니다.
     */
    override fun handleRequest(senderSession: WebSocketSession, request: WriteMessage) {
        val senderUserId = senderSession.attributes[IdKey.USER_ID.value] as? UserId ?: return
        val channelId = request.channelId
        val content = request.content
        val senderUsername = userService.getUsername(senderUserId) ?: "unknown"

        messageService.sendMessage(
            senderUserId,
            content,
            channelId,
            MessageNotification(channelId, senderUsername, content)
        )
    }
}
